### DISCUSSION TITLE
How to reduce memory consumption of uBO? It keeps crashing

### DISCUSSION THREAD

**Discussion by XCanG** [uBlockOrigin/uBlock-issues#3368] [Q&A] [answered]
I have an issue that is somewhat consistent, but not appear very often (once in a week). uBO keeps consuming a lot of RAM until it reaches about 10 GB and then extension crashed because "Out of memory". I can keep track of the memory in Google Chrome internal task manager (Shift+Esc). To me it feels like uBO keeps some kind of logs or something that accumulate over time and the more it blocks, the faster it crashes. I find out that using site like YouTube that have a lot of ads (like 500-1000 per tab) make this accumulation of RAM faster.

I don't know what else to add, but I didn't find any options that can reduce RAM usage or could clear current RAM.
<details><summary>Screenshot from task manager</summary>

![image](https://github.com/user-attachments/assets/2495b5ee-8f79-438a-a79f-bb225e6c9137)

</details>

Specs:
* Windows 10 x64
* Google Chrome v126.0.6478.127
* uBlock Origin 1.59.0

**Comment by garry-ut99** [comment]
Looking at uBO documentation, it seems neither high RAM usage nor crashes, are considered uBO's fault:
- https://github.com/gorhill/uBlock/wiki/Myth:-uBlock-consumes-several-or-several-dozen-GB-of-RAM
- https://github.com/gorhill/uBlock/wiki/Can-uBO-crash-a-browser%3F

but I'm not sure about your case, it would be good if you could make sure that your issue is not just about virtual RAM.

**Reply by XCanG** [reply]
The crash was for extension, but it can be reloaded again in extension menu. But it's process, that run extension uses this amount of RAM. While I use other extensions and they do not consume this much RAM and don't crash.

**Comment by gorhill** [comment]
> Google Chrome v126.0.0.0.0

Is this really the exact version of your Google Chrome? If not, can you provide the exact version?

**Reply by XCanG** [reply]
Sorry for long response. I forget to edit in exact version, it's `126.0.6478.127`

**Reply by gorhill** [reply]
In case this could be related to extension storage corruption, have you tried removing then reinstalling uBO?

Also, anything installed in your computer which potentially hooks into processes, such as anti-viruses or others?

**Reply by XCanG** [reply]
When it crashes it says "reload extension", but it looks like that it reinstall it rather than just reload, so I guess this restore storage.

I don't use anti-viruses, but I use [GoodbyeDPI](https://github.com/ValdikSS/GoodbyeDPI) to bypass some of the ISP/government blocks, it modify traffic and youtube is the one of the sites that was recently affected, so if it somehow related to that, then may be, but I can't work without it, unless I use VPN as alternative, though it's not perfect.

**Reply by gorhill** [reply]
Reloading is not the same as reinstall. Removing the extension will remove all storages associated with it, so this step is key if the issue is extension storage corruption.

Also can you share the troubleshooting information from dashboard => Support => Troubleshooting information => copy/paste content here.

**Reply by XCanG** [reply]
```yml
uBlock Origin: 1.59.0
Chromium: 126
filterset (summary):
 network: 177665
 cosmetic: 132551
 scriptlet: 34050
 html: 0
listset (total-discarded, last-updated):
 added:
  JPN-1: 15173-30, 1d.12h.35m
  KOR-1: 5083-47, 34m
  adguard-spyware-url: 1695-7, 1d.12h.34m
  fanboy-cookiemonster: 52088-235, 12h.34m
  fanboy-thirdparty_social: 68-0, 7d.13h.29m
  ublock-annoyances: 6003-13, 1h.34m Δ
  ublock-cookies-easylist: 1720-9, 2d.12h.35m
 default:
  user-filters: 28-0, never
  RUS-0: 40721-105, 1h.34m Δ
  easylist: 85379-651, 1h.34m Δ
  easyprivacy: 53061-703, 1h.34m Δ
  plowe-0: 3543-0, 3d.13h.36m
  ublock-badware: 11213-6, 1h.34m Δ
  ublock-filters: 40030-305, 1h.34m Δ
  ublock-privacy: 1164-144, 1h.34m Δ
  ublock-quick-fixes: 142-0, 1h.34m Δ
  ublock-unbreak: 2469-1, 1h.34m Δ
  urlhaus-1: 27100-0, 12h.34m
filterset (user): [array of 30 redacted]
trustedset:
 added: [array of 7 redacted]
hostRuleset:
 added: [array of 3 redacted]
userSettings:
 advancedUserEnabled: true
 cloudStorageEnabled: true
 suspendUntilListsAreLoaded: true
hiddenSettings: [none]
supportStats:
 allReadyAfter: 1634 ms
 maxAssetCacheWait: 243 ms
 cacheBackend: indexedDB
```

**Reply by gorhill** [reply]
Are there any uBO-related pages opened for a long time when you experience the high-memory usage, either the dashboard, logger, or any pages owned by uBO?

**Reply by XCanG** [reply]
If only popup on navbar, but no other pages used.

**Reply by XCanG** [reply]
I've check todays RAM consumption and it's already increased by a lot, 1.6 GB:
https://github.com/user-attachments/assets/f54ecf18-e618-496b-b6b1-d4a2fc1e4042

**Reply by gorhill** [reply]
Not much I can do with a heavily cropped GIF, this just begs for more information about all the details of what is going on in your browser. 

Is uBO the only process present in your browser? Have you tried to find out if the issue is related to a specific webpage from a specific site being opened for the issue to appear? Have you tried to narrow down if another extension must be enabled for the issue to appear. The possibilities are countless, so the first step is to always minimize all external factors. Once you do this, opening uBO's dev tools => Memory => Collect garbage; does this lower memory usage? If not, clicking "Take snapshot", saving and sharing the result might help me to find out what is being allocated in uBO to cause this abnormal memory usage.

**Reply by XCanG** [reply]
It's a video, other than it it's other tabs within a browser, I just don't see the reason to show them, just a day after it's climbed up to 2.4 GB already, here is full screenshot of the task manager if you really need it: 
![image](https://github.com/user-attachments/assets/893b085c-56eb-4b0e-a561-f26dec577303)

> Is uBO the only process present in your browser?

Don't quite understand this question, Chrome always create subprocesses for other tabs and I have other extensions as well, but other extension processes at much took 200 MB or less and don't climb over time until process crashed, so it's only uBO that climb and climb until it run out of memory and crash.

> Have you tried to find out if the issue is related to a specific webpage from a specific site being opened for the issue to appear?

It hard to tell, I always have tabs on various websites, but it is YouTube that have most tabs, because I keep some music tabs at hand. I'm working remotely, so I always stick to listening some music at background, that's why I have many tabs from YouTube. And YT is known to have a lot of ads, when I left tab for a while and check it later, uBO shows 1k+ ads that being blocked, so my assumption it is that YouTube that keep showing ads in bg, uBO that keep blocking them and keep some logs or stuff? then it just run out of RAM and crash.

> Have you tried to narrow down if another extension must be enabled for the issue to appear.

I have not, because they aren't related to ads, should be ignored and don't mess up with other addons and don't take much RAM.

**Reply by stephenhawk8054** [reply]
As always, don't assume extensions can't conflict with each other. We are not those extensions' devs, you are not those extensions' devs. We cannot assure 100% that there would be no conflicts.

The only ways to investigate what is going wrong with an extension is running that extension alone. That would help the investigations a lot.

**Reply by XCanG** [reply]
> Once you do this, opening uBO's dev tools

Where is this dev tools again? Or are you asking to open Ctrl+Shift+I devtools on addon page?

**Reply by gorhill** [reply]
> Where is this dev tools again?

Navigate to `chrome://extensions/`.

Enable "Developer mode":
![image](https://github.com/user-attachments/assets/12caf029-4cc4-48d6-b31b-f9245ddc2fbd)

 This will add links to background process to each extension, click the one for uBO, the "background.html" link:
![image](https://github.com/user-attachments/assets/b4215ce7-9140-4001-a481-16388411a51c)

This will open dev tools for uBO's process. Click "memory" pane:
![image](https://github.com/user-attachments/assets/5e8d2666-c53a-4a55-9fa0-e6b5babf238e)

It shows how much memory is used by uBO's "JavaScript VM instance".

Post what it says here when you suffer the undue memory usage issue.

Then click the "brush" to force a "Collect garbage", will will force the browser to claim memory which is not used by uBO. Wait ~5 seconds after clicking and post what it says.

If memory usage uBO still shows as high, click the "Take snapshot" button, wait for it to complete, than save results using the "Save profile" button. You can then share the profile data. It may contain personal information, in which case tell me if you want to send it directly to me instead of publishing the file here.

**Reply by XCanG** [reply]
> Post what it says here when you suffer the undue memory usage issue.

![image](https://github.com/user-attachments/assets/42e17f67-bef2-4d5d-92b2-33fb658add4e)

Hm, it shows here only 25.1 MB.

When I click "Collect garbage" button nothing happens, I wait for 5 minutes, if there supposed to be some popup then I don't see any, however one thing was changed in task manager: RAM usage dropped from 2.4 GB to 1.6 GB.

If I'm gonna send snapshot, then definitely want to send it directly.

**Reply by gorhill** [reply]
Also please post your troubleshooting information from SUpport pane in uBO's dashboard.

>  I wait for 5 minutes

You don't have to wait minutes, just 5 seconds is ok, it's just to be sure the browser complete the garbage collection task. The fact that the task manager drops ~800MB suggest the browser is not reclaiming memory which has been freed by uBO. Typically uBO minimizes memory allocation, and if there happen to be more than is expected, this could be caused by another extension causing uBO to work more than it would normally do, hence why disabling all other extensions is key to rule out external factor, no assumption should be made that other extensions can't be the root cause, all extensions work in the same DOM and can affect each others this way.

**Reply by XCanG** [reply]
> Also please post your troubleshooting information from SUpport pane in uBO's dashboard.

```yml
uBlock Origin: 1.59.0
Chromium: 126
filterset (summary):
 network: 175902
 cosmetic: 132686
 scriptlet: 34117
 html: 0
listset (total-discarded, last-updated):
 added:
  JPN-1: 15183-30, 11h.17m
  KOR-1: 5085-47, 10h.17m
  adguard-spyware-url: 1713-7, 11h.17m
  fanboy-cookiemonster: 52088-237, 4d.11h.22m
  fanboy-thirdparty_social: 68-0, 2d.11h.21m
  ublock-annoyances: 5970-12, 1h.17m Δ
  ublock-cookies-easylist: 1720-9, 6d.11h.24m
 default:
  user-filters: 28-0, never
  RUS-0: 40710-105, 17m Δ
  easylist: 85363-649, 1h.17m Δ
  easyprivacy: 53075-703, 1h.17m Δ
  plowe-0: 3543-0, 7d.12h.25m
  ublock-badware: 11271-6, 1h.17m Δ
  ublock-filters: 40233-304, 1h.17m Δ
  ublock-privacy: 1192-140, 1h.17m Δ
  ublock-quick-fixes: 153-0, 1h.17m Δ
  ublock-unbreak: 2464-2, 1h.17m Δ
  urlhaus-1: 25255-0, 11h.17m
filterset (user): [array of 30 redacted]
trustedset:
 added: [array of 7 redacted]
hostRuleset:
 added: [array of 3 redacted]
userSettings:
 advancedUserEnabled: true
 cloudStorageEnabled: true
 suspendUntilListsAreLoaded: true
hiddenSettings: [none]
supportStats:
 allReadyAfter: 1634 ms
 maxAssetCacheWait: 243 ms
 cacheBackend: indexedDB
```

**Reply by XCanG** [reply]
I also check Console tab in devtools and find that there are 570 errors of same kind:

![image](https://github.com/user-attachments/assets/33df16e3-081b-40cc-846f-0bf356fb2493)

Could it be related?

**Reply by garry-ut99** [reply]
> > > gorhill: Have you tried to narrow down if another extension must be enabled for the issue to appear.
> >
> >  XCanG: I have not, because they aren't related to ads, should be ignored and don't mess up with other addons and don't take much RAM.
>
> stephenhawk8054 : don't assume extensions can't conflict with each other. (...)
> gorhill: no assumption should be made that other extensions can't be the root cause (...)


Yes, even theoretically unrelated extensions might affect other extensions and uBO, here is a recent example where an other extension - Bitwarden password manager affected uBO badly: https://github.com/uBlockOrigin/uBlock-issues/issues/3217#issuecomment-2106253846 + the next 3 comments.

**Reply by XCanG** [reply]
> a recent example where an other extension - Bitwarden password manager affected uBO badly

Ohh, didn't know that, I also use Bitwarden.

**Reply by gorhill** [reply, marked as answer]
> there are 570 errors of same kind

This is fixed in dev build with <https://github.com/gorhill/uBlock/commit/1b464f75cc6b036d520c04944c5c522b5671b72d>.

> Could it be related?

No clue. You could try the dev build to see if this changes something.
